### DISCUSSION TITLE
There will be problems when using verb, font, xelatex and CJK character in the begin of cell at the same time.

### DISCUSSION THREAD

**Discussion by note286** [lvjr/tabularray#370] [Q&A]
# Not Working MWE

```latex
% xelatex
\documentclass{ctexart}
\usepackage{tabularray}
\begin{document}
\begin{tblr}{colspec={c},row{1}={font=\bfseries},verb=true}
测test\\
\verb|a\b\b|\\
\end{tblr}
\end{document}
```

```
(c:/texlive/2022/texmf-dist/tex/latex/tabularray/tabularray.sty) (./1.aux)
! Undefined control sequence.
l.1 \bfseries\selectfont测test
                             \empty
l.7 \end
        {tblr}
?
```

# Working MWE

change `xelatex` to `pdflatex`

```latex
% pdflatex
\documentclass{ctexart}
\usepackage{tabularray}
\begin{document}
\begin{tblr}{colspec={c},row{1}={font=\bfseries},verb=true}
测test\\
\verb|a\b\b|\\
\end{tblr}
\end{document}
```

remove the CJK character in the begin of cell

```latex
% xelatex
\documentclass{ctexart}
\usepackage{tabularray}
\begin{document}
\begin{tblr}{colspec={c},row{1}={font=\bfseries},verb=true}
te测st测\\
\verb|a\b\b|\\
\end{tblr}
\end{document}
```

change `font` key to `cmd` key

```latex
% xelatex
\documentclass{ctexart}
\usepackage{tabularray}
\begin{document}
\begin{tblr}{colspec={c},row{1}={cmd=\bfseries},verb=true}
测test\\
\verb|a\b\b|\\
\end{tblr}
\end{document}
```

do not use `verb`

```latex
% xelatex
\documentclass{ctexart}
\usepackage{tabularray}
\begin{document}
\begin{tblr}{colspec={c},row{1}={font=\bfseries}}
测test\\
% \verb|a\b\b|\\
\end{tblr}
\end{document}
```

**Comment by muzimuzhi** [comment]
A re-tokenization problem. XeLaTeX (and also LuaLaTeX) treats (most) CJK characters the same as a-zA-Z, thus after rescanning the table content, XeTeX thinks `\selectfont测` is a single control sequence, rather than a control sequence `\selectfont` followed by a character `测`。

Some workarounds:
```tex
% !TeX TS-program = xelatex
\documentclass{ctexart}
\usepackage{tabularray}
\usepackage{codehigh}
\usepackage{fvextra}

\begin{document}
% "{}" trick
\begin{tblr}{colspec={c},row{1}={font=\bfseries},verb=true,hlines,vlines}
  {}测test\\
  \verb|a\b\b|\\
\end{tblr}

% \fakeverb from codehigh package, which shares the same author with tabularray
\begin{tblr}{colspec={c},row{1}={font=\bfseries},hlines,vlines}
  测test\\
  \fakeverb{a\\b\\b}\\
\end{tblr}

% \EscVerb from fvextra package
\begin{tblr}{colspec={c},row{1}={font=\bfseries},hlines,vlines}
  测test\\
  \EscVerb{a\\b\\b}\\
\end{tblr}
\end{document}
```
![image](https://user-images.githubusercontent.com/6376638/221390951-d8f6b9f0-85d3-45d2-a909-bfbdb85ab9a7.png)

Similar though not identical re-tokenization issue
- https://github.com/lvjr/tabularray/issues/358

**Reply by muzimuzhi** [reply]
The latter two workarounds are better than the first, cause they both avoids the re-tokenization, which may be the source of all kinds of problems.

**Reply by note286** [reply]
Thank you. I'm clear.

**Reply by lvjr** [reply]
This is a duplicate of issue #294. I am planning to mark `verb` option as obsolete in a future version and remove it eventually some years later.

**Reply by lvjr** [reply]
In fact `\fakeverb` is simpler than `\EscVerb` because you can write `\fakeverb{a\b\b}`.

**Reply by muzimuzhi** [reply]
btw, the l3regex-based implementations are so slow.

**Reply by Sophanatprime** [reply]
Another implement would be to use `\tl_analysis_map_inline:n`, which is pretty fast:
```
\tl_const:Nn \c__cus_verbatimize_macros_tl { \#\%\\\ \{\}\^ }
\NewDocumentCommand \FakeVerb { m }
  {
    \group_begin:
    \ttfamily \frenchspacing
    \tl_analysis_map_inline:nn {#1}
      { 
        \int_compare:nNnTF {##2} = { -\c_one_int }
          {
            \exp_args:NNo \tl_if_in:NnTF \c__cus_verbatimize_macros_tl {##1} 
                { \exp_after:wN \cs_to_str:N ##1 } { \exp_after:wN \token_to_str:N ##1 }
          }
          { \exp_after:wN \token_to_str:N ##1 }
      }
    \group_end:
  }
```

**Reply by lvjr** [reply]
Yes, `l3regex` is quite slow. And I have to improve the speeds of `codehigh` and `tabularray` packages by using less `l3regex` and less loops. As a first step I need to write some benchmark tests.

**Comment by lvjr** [comment]
Since version 2023A, `verb` option was obsolete. Please use `\fakeverb` command from `codehigh` package instead.
